consensus/cryptarchia-sync/src/downloader.rs

```diff
@@ -0,0 +1,133 @@
+use futures::{future::BoxFuture, stream, stream::BoxStream, StreamExt as _};
+use libp2p::{PeerId, Stream as Libp2pStream};
+use libp2p_stream::Control;
+use tokio::sync::mpsc::Sender;
+use tracing::error;
+
+use crate::{
+    behaviour::{BlocksResponse, RequestStream, ResponseKind, TipResponse},
+    errors::ChainSyncError,
+    messages::{DownloadBlocksRequest, DownloadBlocksResponse, GetTipResponse, RequestMessage},
+    packing::unpack_from_reader,
+    utils,
+    utils::{open_stream, send_message},
+    ChainSyncErrorKind,
+};
+
+pub struct DownloadBlocksTask;
+
+impl DownloadBlocksTask {
+    pub async fn send_tip_request(
+        peer_id: PeerId,
+        control: &mut Control,
+        reply_sender: Sender<BoxStream<'static, Result<ResponseKind, ChainSyncError>>>,
+    ) -> Result<
+        (
+            RequestStream,
+            Sender<BoxStream<'static, Result<ResponseKind, ChainSyncError>>>,
+        ),
+        ChainSyncError,
+    > {
+        let mut stream = open_stream(peer_id, control).await?;
+
+        let tip_request = RequestMessage::GetTip;
+        send_message(peer_id, &mut stream, &tip_request).await?;
+
+        let request_stream = RequestStream::new(peer_id, stream, tip_request);
+        Ok((request_stream, reply_sender))
+    }
+
+    pub async fn send_download_request(
+        peer_id: PeerId,
+        mut control: Control,
+        request: DownloadBlocksRequest,
+        reply_sender: Sender<BoxStream<'static, Result<ResponseKind, ChainSyncError>>>,
```

Comment:
Same here, why not `BlockResponse`. Who then unpacks this? the consumer service is probably expecting `Blocks`